core/services/vrf/v2/listener_v2_log_listener_test.go

```diff
@@ -0,0 +1,688 @@
+	})
+
+	// Block 3 to finalityDepth. Ensure we have finality number of blocks
+	for i := 1; i < int(finalityDepth); i++ {
+		th.Client.Commit()
+	}
+
+	// Make a VRF request without fulfilling it
+	_, err2 := th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+		[32]byte(th.Listener.job.VRFSpec.PublicKey.MustHash().Bytes()),
+		big.NewInt(1), big.NewInt(105), 1, 10, 10000, 2, th.Owner.From)
+	require.NoError(t, err2)
+	th.Client.Commit()
+
+	// Emit some logs in blocks to make the VRF req and fulfillment older than finalityDepth from latestBlock
+	numBlocks := 5
+	th.Client.Commit()
+	for i := 0; i < numBlocks; i++ {
+		_, err1 := th.Emitter.EmitLog1(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		_, err1 = th.Emitter.EmitLog2(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 1 (VRF req block) + 5 (EmitLog blocks) = 10
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(th.Ctx))
+
+	latestBlock := finalityDepth + int64(numBlocks) + 1
+	require.NoError(t, th.LogPoller.Replay(th.Ctx, latestBlock))
+
+	// initializeLastProcessedBlock must return the unfulfilled VRF
+	// request block number (5) instead of finalizedBlockNumber (8)
+	lastProcessedBlock, err := th.Listener.initializeLastProcessedBlock(th.Ctx)
+	require.Nil(t, err)
+	require.Equal(t, int64(5), lastProcessedBlock)
+}
+
+func TestInitProcessedBlock_SomeUnfulfilledVRFReqs(t *testing.T) {
+	t.Parallel()
+
+	finalityDepth := int64(3)
+	th := SetupTH(t, false, finalityDepth, 3, 2, 1000, func(mockChain *evmmocks.Chain, curTH *TestHarness) {
+		mockChain.On("ID").Return(curTH.ChainID)
+		mockChain.On("LogPoller").Return(curTH.LogPoller)
+	})
+
+	// Block 3 to finalityDepth. Ensure we have finality number of blocks
+	for i := 1; i < int(finalityDepth); i++ {
+		th.Client.Commit()
+	}
+
+	// Emit some logs in blocks with VRF reqs interspersed
+	// No fulfillment for any VRF requests
+	numBlocks := 5
+	th.Client.Commit()
+	for i := 0; i < numBlocks; i++ {
+		_, err1 := th.Emitter.EmitLog1(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		_, err1 = th.Emitter.EmitLog2(th.Owner, []*big.Int{big.NewInt(int64(i))})
+		require.NoError(t, err1)
+		th.Client.Commit()
+
+		// Create 2 blocks with VRF requests in each iteration
+		_, err2 := th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+			[32]byte(th.Listener.job.VRFSpec.PublicKey.MustHash().Bytes()),
+			big.NewInt(int64(2*i)), big.NewInt(105), 1, 10, 10000, 2, th.Owner.From)
+		require.NoError(t, err2)
+		th.Client.Commit()
+		_, err2 = th.VRFLogEmitter.EmitRandomWordsRequested(th.Owner,
+			[32]byte(th.Listener.job.VRFSpec.PublicKey.MustHash().Bytes()),
+			big.NewInt(int64(2*i+1)), big.NewInt(106), 1, 10, 10000, 2, th.Owner.From)
```

Comment:
Extract important fields into named vars

Reply:
Used named vars for important parameters